devices/Button/ButtonBase.cs

```diff
@@ -0,0 +1,136 @@
+
+        private int _doublePressMs;
+        private int _holdingMs;
+
+        private ButtonHoldingState _holdingState = ButtonHoldingState.Completed;
+
+        private long _lastPress = DateTime.MinValue.Ticks;
+        private Timer _holdingTimer;
+
+        public delegate void ButtonPressedDelegate(object sender, EventArgs e);
+        public delegate void ButtonHoldingDelegate(object sender, ButtonHoldingEventArgs e);
+
+        public event ButtonPressedDelegate ButtonUp;
+        public event ButtonPressedDelegate ButtonDown;
+        public event ButtonPressedDelegate Press;
+        public event ButtonPressedDelegate DoublePress;
+        public event ButtonHoldingDelegate Holding;
+
+        public bool IsHoldingEnabled { get; set; } = false;
+        public bool IsDoublePressEnabled { get; set; } = false;
+        public bool IsPressed { get; set; } = false;
+
+        /// <summary>
+        /// Initialization of the button.
+        /// </summary>
+        public ButtonBase(int doublePressMs = DEFAULT_DOUBLE_PRESS_TICKS, int holdingMs = DEFAULT_HOLDING_MS)
+        {
+            _doublePressMs = doublePressMs;
+            _holdingMs = holdingMs;
+        }
+
+        /// <summary>
+        /// Handler for pressing the button.
+        /// </summary>
+        protected void HandleButtonPressed()
+        {
+            IsPressed = true;
+
+            ButtonDown?.Invoke(this, new EventArgs());
+
+            if (IsHoldingEnabled)
+            {
+                _holdingTimer = new Timer(StartHoldingHandler, null, _holdingMs, Timeout.Infinite);
+            }
+        }
+
+        /// <summary>
+        /// Handler for releasing the button.
+        /// </summary>
+        protected void HandleButtonReleased()
+        {
+            _holdingTimer?.Dispose();
+            _holdingTimer = null;
+
+            IsPressed = false;
+
+            ButtonUp?.Invoke(this, new EventArgs());
+            Press?.Invoke(this, new EventArgs());
+
+            if (IsHoldingEnabled && _holdingState == ButtonHoldingState.Started)
+            {
+                _holdingState = ButtonHoldingState.Completed;
+                Holding?.Invoke(this, new ButtonHoldingEventArgs { HoldingState = ButtonHoldingState.Completed });
+            }
+
+            if (IsDoublePressEnabled)
+            {
+                if (_lastPress == DateTime.MinValue.Ticks)
+                {
+                    _lastPress = DateTime.UtcNow.Ticks;
+                }
+                else
+                {
+                    if (DateTime.UtcNow.Ticks - _lastPress <= _doublePressMs)
+                    {
+                        DoublePress.Invoke(this, new EventArgs());
+                    }
+                    _lastPress = DateTime.MinValue.Ticks;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handler for holding the button.
+        /// </summary>
+        private void StartHoldingHandler(object state)
```

Comment:
This is necessary because otherwise in line 201
`_holdingTimer = new Timer(StartHoldingHandler, null, _holdingMs, Timeout.Infinite);
`
It will complain that it cannot convert StartHoldingHandler to 'method group' to 'timercallback'